# Исследование надежности заемщиков.

**Цель проекта:** выснить, влияет ли семейное положение, количество детей, уровень дохода и цели кредита на срок возврата долга по кредиту в срок в банк.

### Шаг 1. Обзор данных

In [1]:
import pandas as pd

In [2]:
#чтение датасета и сохранение его в переменную data
try:
    data = pd.read_csv('D:\_THIS IS MY FOLDER\Я.Практикум, аналитика\data.csv')
except:
    data = pd.read_csv('/datasets/data.csv')
    

In [3]:
data.head(10) #вывела на экран первые 10 строк таблицы, чтобы познакомиться с данными

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [4]:
data.info() #получение общей информаии о таблице с данными

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице всего 21525 строк. Вижу из общей информации, что в столбцах 'days_employed' и 'total_income' есть пропущенные значения (потому что 19351 не равно '21525).<br> 
С наименованиями столбцов всё в порядке.<br>

Проверю, как выглядят пропущенные значения в столбцах days_employed' и 'total_income'  (это просто пропуски, Nan, None или какие-то другие символы?)

In [5]:
data[data['days_employed'].isna()].head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

In [6]:
data[data['total_income'].isna()].head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Узнаю, какую долю составляют пропущенные значения в каждом из столбцов с пропусками

In [7]:
print (('Доля пропущенных значений в столбце "days_employed": {:.0%}'.format(data['days_employed'].isna().mean())))

Доля пропущенных значений в столбце "days_employed": 10%


In [8]:
print (('Доля пропущенных значений в столбце "total_income": {:.0%}'.format(data['total_income'].isna().mean())))

Доля пропущенных значений в столбце "total_income": 10%


Пропущенные значения - это NaN. Данных нет. <br>
Nan - это тип данных float. Значит, с этими значениями можно проводить математические операции. Поэтому встает выбор: заполнить пропущенные значения медианой или средним значением или вовсе удалить. <br>
Судя по доле пропущенных значений и по выводимым значениям, в тех строках, где есть пропуск в столбце 'days_employed', в тех же строках есть пропуск в столбце 'total_income'. Значит, данные отсутствуют по каким-то конкретным причинам. Пропуск может быть на основании другой имеющейся информации внаборе данных (образование, пол, стаж работы) - Missing At Random. Для таких пропущенных данных неприменимо игнорирование или удаление. Значит, будем заполнять данные средним значением или медианой. 

Посмотрим на разброс значений в столбцах 

In [9]:
data['days_employed'].abs().describe()

count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

Вижу, что в этом столбце есть такие странные значения как 354500.415854 и 401755.400475. То есть как будто стаж человека около 1000 лет?. Этого не может быть. Предположу, что максимально возможный трудовой стаж для человека - 60 лет (если он начал работать с 20 лет, то тогда закончит в 80). Это 60*360 =21600 дней. Тогда нам нужно избавиться от всех значений, которые больше 21600 и заменить их на NaN. 

In [10]:
data['total_income'].abs().describe()

count    1.935100e+04
mean     1.674223e+05
std      1.029716e+05
min      2.066726e+04
25%      1.030532e+05
50%      1.450179e+05
75%      2.034351e+05
max      2.265604e+06
Name: total_income, dtype: float64

Лучше использовать медиану, т.к. в данных значения имеют большой разброс. Такие данные сместят среднее.

**Выводы** <br>
<ol>
    
<li>В каждй строке таблицы информация о клиенте банка: количество детей, стаж работы, возраст, образование, семейное положение, пол, статус занятости, доход и цель кредита.<br>
<li>Предварительно можно сказать, что данных достаточно для анализа. Однако в данных есть пропуски, разные регистры строковых объектов в заполненных ячейках, отрицательные значения там, где должны быть положительные и аномально отклоняющиеся значения, которые не могут быть в природе <br>
  <li>Чтобы двигаться дальше, нужно сделать предобработку данных.<br>
      </ol>

### Шаг 2.1 Заполнение пропусков

Прежде чем считать медиану и заполнять ею пропущенные значения, избавлюсь от аномалий в столбце ['days_employed'] Я заполню их числовыми значениями NaN

In [11]:
data.loc[data['days_employed'] > 21600, 'days_employed'] = pd.NA

In [12]:
#присваиваю пропущенным значениям в столбцах соответствующие медианные значения в зависимости от типа занятости
data['days_employed'] = data['days_employed'].fillna(data.groupby('income_type')['total_income'].transform("median"))
data['total_income'] = data['total_income'].fillna(data.groupby('income_type')['total_income'].transform("median"))

In [13]:
#проверяю, остались ли ещё в таблице пропущенные значения
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Шаг 2.2 Проверка данных на аномалии и исправления.

Как мы выяснили ранее, в столбце `days_employed` есть отрицательные значения. Заменю их на те же самые, только со знаком "+".

In [14]:
data['days_employed'] = data['days_employed'].abs()

In [15]:
#выведу первые 10 строк таблицы, чтобы убедиться, что отрицательных чисел в столбце больше нет
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  118514.486412         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [16]:
#и на всякий случай отсортирую числа в столбце 'days_employed' 
#по возрастанию, чтобы 100%-но убедиться, что отрицательных чисел нет
data['days_employed'].value_counts(ascending=False)

118514.486412    3856
142594.396847    1105
172357.950966     508
150447.935283     147
131339.751676       2
                 ... 
445.286433          1
1133.987382         1
5730.178239         1
429.448441          1
1984.507589         1
Name: days_employed, Length: 15912, dtype: int64

In [17]:
#проверю, сколько детей может быть у людей
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

In [18]:
#-1 ребенок не может быть, поэтому заменим -1 на 1
#20 детей в семье может быть. в нашей стране аж 76 таких семей:
data['children'] = data['children'].abs()

### Шаг 2.3. Изменение типов данных.

In [19]:
#заменим значения в столбце total_income на целочисленный
data['total_income'] = data['total_income'].astype('int') 

### Шаг 2.4. Удаление дубликатов.

In [20]:
#проверим, есть ли в наборе данных дублирующиеся строки:
data.duplicated().sum()

54

In [21]:
#удалим дубликаты с обновлением индексации столбцов и удалением старых
data = data.drop_duplicates().reset_index(drop=True)

In [22]:
#посмотрим, какие виды образования могут быть у людей
data['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

Одни и те же значения записаны по-разному: с разными регистрами букв. Приведём все значения к единому виду и оставим только 5 возможных:<ul>
<li> высшее
    <li>среднее
        <li>неоконченное высшее 
            <li>начальное 
                <li>ученая степень.

In [23]:
#посмотрим, какие цели займов могут быть у людей
data['purpose'].value_counts()

свадьба                                   793
на проведение свадьбы                     773
сыграть свадьбу                           769
операции с недвижимостью                  675
покупка коммерческой недвижимости         662
покупка жилья для сдачи                   652
операции с жильем                         652
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          625
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

Одни и те же цели сформулированы по-разному. Оставим только несколько возможных целей:<br>
<ol>
<li>'проведение свадьбы'
<li>'получение образования'
    <li>'операции с недвижимостью'
            <li>'операции с автомобилем'
                </ol>
                

Проверим наличие дубликатов в других столбцах с типом данных object 

In [24]:
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [25]:
data['income_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

In [26]:
data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

В столбцах `gender`, `family_status`,`income_type` нет неявных дубликатов

Возможная причина появления дубликатов данных в столбцах `purpose` и `education`: скорее всего значения этих столбцов люди заполняли вручную (например, в форме обратной связи на сайте), а не выбирали из предложенного списка. Поэтому каждый человек сформулировал значение по-своему

In [27]:
#приведем значения в столбце 'education' к одному регистру
data['education'] = data['education'].str.lower()

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [28]:
#создание датафрейма edication со столбцами education_id и education из исходного датафрейма
education = data[['education','education_id']]

#создание датафрейма family со столбцами family_status_id и family_status из исходного датафрейма
family = data[['family_status','family_status_id']]

#удалим из исходного датафрейма столбец education
data.pop('education')


0         высшее
1        среднее
2        среднее
3        среднее
4        среднее
          ...   
21466    среднее
21467    среднее
21468    среднее
21469    среднее
21470    среднее
Name: education, Length: 21471, dtype: object

In [29]:
#удалим из исходного датафрейма столбец family_status
data.pop('family_status')

0         женат / замужем
1         женат / замужем
2         женат / замужем
3         женат / замужем
4        гражданский брак
               ...       
21466    гражданский брак
21467     женат / замужем
21468    гражданский брак
21469     женат / замужем
21470     женат / замужем
Name: family_status, Length: 21471, dtype: object

In [30]:
#проверю, что столбцы удалены
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  118514.486412         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

### Шаг 2.6. Категоризация дохода.

In [31]:
# функция income_categoty, которая присваивает категорию в зависимости от дохода
# если параметр дохода в диапазоне 0–30000 — категория 'E'; 30001–50000 — 'D'; 50001–200000 — 'C'; 200001–1000000 — 'B';
# 1000001 и выше — 'A'
def income_categoty(row):
    if 0 <= row <= 30000:
        return 'E'
    if 30001 <= row <= 50000:
        return 'D'
    if 50001 <= row <= 200000:
        return 'C'
    if 200001 <= row <= 1000000:
        return 'B'
    return 'A'
#проверяю корректную работу функции:
income_categoty(300000)

'B'

In [32]:
#добавляю в датафрейм data столбец total_income_category, 
#где будут храниться результаты применения функции income_categoty
data['total_income_category'] = data['total_income'].apply(income_categoty)
data

children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028         42             0                 0   
1             1    4024.803754         36             1                 0   
2             0    5623.422610         33             1                 0   
3             3    4124.747207         32             1                 0   
4             0  118514.486412         53             1                 1   
...         ...            ...        ...           ...               ...   
21466         1    4529.316663         43             1                 1   
21467         0  118514.486412         67             1                 0   
21468         1    2113.346888         38             1                 1   
21469         3    3112.481705         38             1                 0   
21470         2    1984.507589         40             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21466      F   компаньон     0        224791             операции с жильем   
21467      F   пенсионер     0        155999          сделка с автомобилем   
21468      M   сотрудник     1         89672                  недвижимость   
21469      M   сотрудник     1        244093  на покупку своего автомобиля   
21470      F   сотрудник     0         82047         на покупку автомобиля   

      total_income_category  
0                         B  
1                         C  
2                         C  
3                         B  
4                         C  
...                     ...  
21466                     B  
21467                     C  
21468                     C  
21469                     B  
21470                     C  

[21471 rows x 11 columns]

### Шаг 2.7. Категоризация целей кредита.

In [33]:
# функция find_purpose_category, которая присваивает категорию в зависимости от цели кредита
# вот такие категории: 'операции с автомобилем','операции с недвижимостью','проведение свадьбы','получение образования'.
def find_purpose_category(row):
    if 'авто' in row:
        return 'операции с автомобилем'
    if 'жи' in row:
        return 'операции с недвижимостью'
    if 'свадьб' in row:
        return 'проведение свадьбы'
    return 'получение образования'
#проверяю работоспособность функции
find_purpose_category('операция с недвижимостью')

'операции с недвижимостью'

In [34]:
#добавляю в датафрейм data столбец purpose_category, 
#где будут храниться результаты применения функции find_purpose_category
data['purpose_category'] = data['purpose'].apply(find_purpose_category)
data

children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028         42             0                 0   
1             1    4024.803754         36             1                 0   
2             0    5623.422610         33             1                 0   
3             3    4124.747207         32             1                 0   
4             0  118514.486412         53             1                 1   
...         ...            ...        ...           ...               ...   
21466         1    4529.316663         43             1                 1   
21467         0  118514.486412         67             1                 0   
21468         1    2113.346888         38             1                 1   
21469         3    3112.481705         38             1                 0   
21470         2    1984.507589         40             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21466      F   компаньон     0        224791             операции с жильем   
21467      F   пенсионер     0        155999          сделка с автомобилем   
21468      M   сотрудник     1         89672                  недвижимость   
21469      M   сотрудник     1        244093  на покупку своего автомобиля   
21470      F   сотрудник     0         82047         на покупку автомобиля   

      total_income_category          purpose_category  
0                         B  операции с недвижимостью  
1                         C    операции с автомобилем  
2                         C  операции с недвижимостью  
3                         B     получение образования  
4                         C        проведение свадьбы  
...                     ...                       ...  
21466                     B  операции с недвижимостью  
21467                     C    операции с автомобилем  
21468                     C  операции с недвижимостью  
21469                     B    операции с автомобилем  
21470                     C    операции с автомобилем  

[21471 rows x 12 columns]

### Ответы на вопросы.

##### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [35]:
#создаю сводную таблицу data_pivot_child, где для каждой из категорий людей с детьми считаю сумму долга и общее количество долга
data_pivot_child = data.pivot_table(index='children', values='debt', aggfunc=['sum', 'count'])

#в этой же сводной таблице создаю столбец 'ratio' и сохраняю в нём значение соотношения 
#людей конкретной категории с долгами к общему количеству людей в этой категории, выраженное в процентах
data_pivot_child['ratio'] = data_pivot_child['sum'] *100 / (data_pivot_child['count'])

#отсортирую эту сводную таблицу по столбцу 'ratio' в порядке возрастания, чтобы было удобнее делать выводы
data_pivot_child.sort_values(by='ratio')


sum  count      ratio
          debt   debt           
children                        
5            0      9   0.000000
0         1063  14107   7.535266
3           27    330   8.181818
1          445   4856   9.163921
2          194   2052   9.454191
4            4     41   9.756098
20           8     76  10.526316

##### Вывод 1:

Минимальная вероятность к просрочке у людей с 5-ю детьми. Максимальная вероятность к просрочке у людей с 20 детьми. <b>НО!</b> если мы примем во внимание, что вероятность того, что кредитозаёмщик будет иметь 20 детей стремится к нулю, то:<br>
   Максимальная вероятость к просрочке у людей с 4-мя детьми. 

##### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [36]:
#создаю сводную таблицу data_pivot_family, где для каждой категории людей, в зависимости от 
#их семейного положения и пола, считаю сумму долга и общее количество долга
data_pivot_family = data.pivot_table(index=['family_status_id', 'gender'] , values='debt', aggfunc=['sum', 'count'])

#обновляю индексацию
data_pivot_family = data_pivot_family.reset_index()

#обновлю названия колонок
data_pivot_family.columns = ['family_status_id', 'sex', 'debt', 'count_of_debt']

#в этой же сводной таблице создаю столбец 'ratio' и сохраняю в нём значение соотношения 
#людей конкретной категории семейного положения с долгами к общему количеству людей в этой категории, выраженное в процентах
data_pivot_family['ratio'] = data_pivot_family['debt'] *100 / (data_pivot_family['count_of_debt'])

#уберу из ранее созданного датафрейма femily дубликаты, чтобы при 
#объединении таблиц у меня не было лишних строчек
family_edited = family.drop_duplicates()

#объединю мою сводную таблицу и ранее созданный датафрейм family по столбцам 'family_status_id', для того,
#чтобы сразу видеть, какому семейному положению соответствует каждый код статуса
data_pivot_family=data_pivot_family.merge(family_edited,on='family_status_id')
data_pivot_family

#выподимую таблицу отсортирую по столбцу 'ratio'
data_pivot_family.sort_values(by='ratio')


family_status_id  sex  debt  count_of_debt      ratio  \
4                  1  XNA     0              1   0.000000   
5                  2    F    52            904   5.752212   
7                  3    F    61            936   6.517094   
9                  4    F   118           1729   6.824754   
0                  0    F   530           7764   6.826378   
2                  1    F   233           2856   8.158263   
1                  0    M   401           4580   8.755459   
8                  3    M    24            259   9.266409   
3                  1    M   155           1306  11.868300   
10                 4    M   156           1081  14.431082   
6                  2    M    11             55  20.000000   

            family_status  
4        гражданский брак  
5          вдовец / вдова  
7               в разводе  
9   Не женат / не замужем  
0         женат / замужем  
2        гражданский брак  
1         женат / замужем  
8               в разводе  
3        гражданский брак  
10  Не женат / не замужем  
6          вдовец / вдова

Минимальная вероятность к просрочке у женщин-вдов. Максимальная вероятность к просрочке у мужчин-вдовцов.

##### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [37]:
#делаю сводную таблицу, где для каждой категории людей, в зависимости от их дохода,
#считаю сумму долгов и общее количество долгов
data_pivot_income = data.pivot_table(index='total_income_category', values='debt', aggfunc=['sum', 'count'])

#считаю соотношение суммы долгов в каждой категории к общему количеству долгов
data_pivot_income ['ratio'] = data_pivot_income['sum'] *100 / data_pivot_income['count']

#вывожу таблицу, отсортированную по возрастанию по столбцу 'ratio'
data_pivot_income.sort_values(by='ratio')

sum  count     ratio
                       debt   debt          
total_income_category                       
D                        21    350  6.000000
B                       356   5042  7.060690
A                         2     25  8.000000
C                      1360  16032  8.483034
E                         2     22  9.090909

Максимальная вероятность просрочки у граждан категории "E" - доход 0–30000, минимальная вероятность просрочки у граждан категории "D" - доход 30001–50000

##### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

In [38]:
#делаю сводную таблицу, где для каждой категории людей, в зависимости от их цели кредита,
#считаю сумму долгов и общее количество долгов
data_pivot_purpose = data.pivot_table(index='purpose_category', values='debt', aggfunc=['sum', 'count'])

#считаю соотношение суммы долгов в каждой категории к общему количеству долгов
data_pivot_purpose['ratio'] = data_pivot_purpose ['sum'] * 100 / data_pivot_purpose ['count']

#вывожу таблицу, отсортированную по возрастанию по столбцу 'ratio'
data_pivot_purpose.sort_values(by='ratio')

sum  count     ratio
                         debt   debt          
purpose_category                              
операции с недвижимостью  782  10814  7.231367
проведение свадьбы        186   2335  7.965739
получение образования     370   4014  9.217738
операции с автомобилем    403   4308  9.354689

Минимальная вероятность просрочки у граждан, которые берут кредит на операции с недвижимостью. Максимальная вероятность просрочки у граждан, которые берут кредит на операции с автомобилем.


## Общий вывод:

Исходя из полученных расчётов, нарисовывается портрет клиента, который имеет наибольшую вероятность просрочки по кредиту: это мужчина-вдовец с 4-мя детьми, с доходом 0–30000, который берёт кредит на операции с автомобилем.